crates/sui-bridge-indexer/src/sui_bridge_indexer.rs

```diff
@@ -0,0 +1,286 @@
+};
+
+/// Persistent layer impl
+#[derive(Clone)]
+pub struct PgBridgePersistent {
+    pool: PgPool,
+}
+
+impl PgBridgePersistent {
+    pub fn new(pool: PgPool) -> Self {
+        Self { pool }
+    }
+}
+
+impl Persistent<ProcessedTxnData> for PgBridgePersistent {
+    fn write(&self, data: Vec<ProcessedTxnData>) -> Result<(), Error> {
+        if data.is_empty() {
+            return Ok(());
+        }
+        let connection = &mut self.pool.get()?;
+        connection.transaction(|conn| {
+            for d in data {
+                match d {
+                    ProcessedTxnData::TokenTransfer(t) => {
+                        diesel::insert_into(token_transfer::table)
+                            .values(&t.to_db())
+                            .on_conflict_do_nothing()
+                            .execute(conn)?;
+
+                        if let Some(d) = t.to_data_maybe() {
+                            diesel::insert_into(token_transfer_data::table)
+                                .values(&d)
+                                .on_conflict_do_nothing()
+                                .execute(conn)?;
+                        }
+                    }
+                    ProcessedTxnData::Error(e) => {
+                        diesel::insert_into(sui_error_transactions::table)
+                            .values(&e.to_db())
+                            .on_conflict_do_nothing()
+                            .execute(conn)?;
+                    }
+                }
+            }
+            Ok(())
+        })
+    }
+}
+
+#[async_trait]
+impl IndexerProgressStore for PgBridgePersistent {
+    async fn load_progress(&self, task_name: String) -> anyhow::Result<u64> {
+        let mut conn = self.pool.get()?;
+        let cp: Option<models::ProgressStore> = dsl::progress_store
+            .find(&task_name)
+            .select(models::ProgressStore::as_select())
+            .first(&mut conn)
+            .optional()?;
+        Ok(cp
+            .ok_or(anyhow!("Cannot found progress for task {task_name}"))?
+            .checkpoint as u64)
+    }
+
+    async fn save_progress(
+        &mut self,
+        task_name: String,
+        checkpoint_number: u64,
+    ) -> anyhow::Result<()> {
+        let mut conn = self.pool.get()?;
+        diesel::insert_into(schema::progress_store::table)
+            .values(&models::ProgressStore {
+                task_name,
+                checkpoint: checkpoint_number as i64,
+                target_checkpoint: i64::MAX,
+                timestamp: None,
+            })
+            .on_conflict(dsl::task_name)
+            .do_update()
+            .set((
+                columns::checkpoint.eq(checkpoint_number as i64),
+                columns::timestamp.eq(now),
+            ))
+            .execute(&mut conn)?;
+        Ok(())
+    }
+
+    fn tasks(&self, prefix: &str) -> Result<Vec<Task>, anyhow::Error> {
+        let mut conn = self.pool.get()?;
+        // get all unfinished tasks
```

Comment:
where is the logic to filter out finished tasks?

Reply:
in the sql query few lines below `.filter(columns::checkpoint.lt(columns::target_checkpoint))`

Reply:
not sure i follow. This is the latest code i'm seeing:

```
        let cp: Vec<models::ProgressStore> = dsl::progress_store
            .filter(columns::task_name.like(format!("{prefix} - %")))
            .filter(columns::checkpoint.lt(columns::target_checkpoint))
            .order_by(columns::checkpoint.desc())
            .load(&mut conn)?;
```

Reply:
not sure i follow. This is the latest code i'm seeing:

```
        let cp: Vec<models::ProgressStore> = dsl::progress_store
            .filter(columns::task_name.like(format!("{prefix} - %")))
            .filter(columns::checkpoint.lt(columns::target_checkpoint))
            .order_by(columns::checkpoint.desc())
            .load(&mut conn)?;
```

Reply:
this query returns all task with name with `prefix` and checkpoint less than target checkpoint, which are all the unfinished tasks

Reply:
yeah this is obvious, now i'm not sure why i asked this question. thanks for explaining!